DropdownMenu v2: tweak styles, add toolbar-specific styles

ciampo opened the pull request:
Part of #50459

<!-- Thanks for contributing to Gutenberg! Please follow the Gutenberg Contributing Guidelines:
https://github.com/WordPress/gutenberg/blob/trunk/CONTRIBUTING.md -->

## What?
<!-- In a few words, what is the PR actually doing? -->

Refine `DropdownMenu` v2 styles:

- updated the border radius of both content wrapper and single items to `2px`;
- updated the box shadow to be the same as the `Popover` component;
- tweak Storybook example to resemble more closely a typical usage of the component;
- added a toolbar-specific `variant` that can be set via only by other components in `@wordpress/components`, which causes:
  - content (and sub-content) wrappers to show a dark border, and remove the drop shadow;
  - separators to also render in the same border color.

**Note: this PR only focuses on tweaking styles on the `DropdownMenu` component. It does NOT make any changes to the `Toolbar` and `ToolbarDropdown` components (and therefore it doesn't impact current dropdowns in the editor). Those changes will be part of a separate PR, and will need some thorough testing around integrating the new `DropdownMenu` component in the editor.**

## Why?
<!-- Why is this PR necessary? What problem is it solving? Reference any existing previous issue(s) or PR(s), but please add a short summary here, too -->

Adapting to design specs.

## How?
<!-- How is your PR addressing the issue at hand? What are the implementation details? -->

The answer is mostly React Context:

- We use the package's [context system](https://github.com/WordPress/gutenberg/blob/trunk/packages/components/CONTRIBUTING.md#context-system) to expose a `variant` property. This property can only be set using the packages' context system, meaning only other components from `@wordpress/components` will be able to trigger this variant. The plan is for this property to be set by the `Toolbar` components
- Internally in the `DropdownMenu` component, we use another instance of React Content to propagate the received `variant` down to all subcomponents. This is to ensure that also the sub-content wrapper and the separator can render accordingly to the chosen variant.

I will add more details about the implementation through inline comments.

## Testing Instructions
<!-- Please include step by step instructions on how to test this PR. -->
<!-- 1. Open a Post or Page. -->
<!-- 2. Insert a Heading Block. -->
<!-- 3. etc. -->

For now, the only way to test the new styles is via Storybook:

- Open the default Storybook example, notice how the border radii have been updated to `2px` for both the content wrappers and each item
- Open the new `Toolbar Variant` Storybook example, and notice how the component renders with darker borders (including nested menus and separators).

## Screenshots or screencast <!-- if applicable -->

https://github.com/WordPress/gutenberg/assets/1083581/f18b7e52-2684-44c1-ad0f-6da7765d875d

ciampo committed:
Update border radius to 2px

ciampo committed:
Add toolbar variant styles to top content wrapper

ciampo committed:
Add toolbar variant styles to sub content wrapper and separators via internal react context

ciampo committed:
Connect `DropdownMenu` to the components context system

ciampo committed:
Add temporary Storybook example

ciampo commented on packages/components/src/dropdown-menu-v2/index.tsx:
```diff
@@ -27,32 +28,47 @@ import type {
+		variant,
+	} = useContextSystem<
+		// Adding `className` to the context type to avoid a TS error
+		DropdownMenuProps & DropdownMenuContext & { className?: string }
```
`useContextSystem` expects `props` to potentially have a `className` prop, but `DropdownMenuProps` doesn't currently expose that prop.

To avoid the type error I'm passing `{ className?: string }` which makes TSC happy without introducing runtime changes, although I'm open to suggestions if folks have other approaches in mind.

tyxla replied:
Is there a reason why we don't expose a `className` as part of `DropdownMenuProps`? We provide it in the v1:

https://github.com/WordPress/gutenberg/blob/22c837b2f93ffed6dc254ba1a721218b2ca8a63a/packages/components/src/dropdown-menu/types.ts#L79

ciampo replied:
For two reasons:

- as a philosophy, we've been trying to expose the least amount of props necessary, and we will add as needed
- currently the component doesn't render a wrapper element — it only renders the trigger (via trigger prop) and the content (via `children` prop). Therefore e're not exactly sure where we should apply the `className`

ciampo commented on packages/components/src/ui/context/index.ts:
```diff
@@ -4,6 +4,7 @@ export {
 } from './context-system-provider';
 export {
 	contextConnect,
+	contextConnectWithoutRef,
```
This function was added in #43611 but was never added to the list of exports

ciampo commented on packages/components/src/dropdown-menu-v2/stories/index.tsx:
```diff
@@ -191,3 +192,19 @@ Default.args = {
+export const ToolbarVariant: ComponentStory< typeof DropdownMenu > = (
+	props
+) => (
+	<ContextSystemProvider value={ toolbarVariantContextValue }>
```
This is how a component in `@wordpress/components` would be able to set this style variant via the package's internal context system

ciampo committed:
Memoize internal context value to avoid extra re-renders

ciampo committed:
CHANGELOG

ciampo committed:
Set the toolbar dropdown menu variant in the toolbar component via the context system

ciampo commented on packages/components/src/toolbar/toolbar/index.tsx:
```diff
@@ -40,12 +51,14 @@ function UnforwardedToolbar(
-			ref={ ref }
-			{ ...props }
-		/>
+		<ContextSystemProvider value={ CONTEXT_SYSTEM_VALUE }>
```
This won't have any effect on dropdowns inside `Toolbar` just yet, since it's still using the legacy `DropdownMenu` component. But it's setting the ground.

tyxla commented on packages/components/src/dropdown-menu-v2/index.tsx:
```diff
@@ -27,32 +33,54 @@ import type {
+		() => ( {
+			variant,
+		} ),
+		[ variant ]
```
It appears that the memoization here could be considered redundant. It will be recalculated only and always when `variant` is updated, and it only contains `variant`, so we can be better off just removing the memoization and directly passing `variant` to `value`. That could help simplify some types as well.

Unless we're intentionally aiming to have more than 1 field provided by the context at some point? I guess if we're going to be with just `variant` for now, we can flatten the object to a single value.

That's almost like using an object with just one property for `useState()`, and changing that object every time, even though the only thing we contain there is just that one value. 

WDYT?

ciampo replied:
I rebased to include changes from https://github.com/WordPress/gutenberg/pull/51046, and the context now includes also the `portalContainer` prop.

I guess that now that we have 2 props and therefore need an object, memoizing the context value could avoid recreating a new object every time the component re-renders, which would have caused consumers of the component to also re-render.

Let me know if you still think that the memoization is extra.

tyxla replied:
Nope, that comment is no longer relevant, I actually reviewed this before #51046 and before seeing that we'll need another context property. 

Feel free to disregard and keep the memoization 👍

tyxla commented on packages/components/src/toolbar/toolbar/index.tsx:
```diff
@@ -16,7 +16,18 @@ import deprecated from '@wordpress/deprecated';
+	ContextSystemProvider,
+} from '../../ui/context';
+
+const CONTEXT_SYSTEM_VALUE = {
```
We may need to add a TODO comment here to address this as we upgrade it to use the v2 dropdown menu component. We could prepend a "TODO" to the comment below I guess.

ciampo replied:
I had already left a note a few lines below, but let me know in case you don't think that's clear enough.

Setting this variable now will also allow us to replace the custom [block settings dropdown ](https://github.com/WordPress/gutenberg/blob/057f64b076b5f9a711e566d724fc6eb5c4228aa0/packages/block-editor/src/components/block-settings-menu/index.js) — using the new `DropdownMenu` component in that file should inherit the context set in the `Toolbar` component

tyxla replied:
My suggestion was to prepend it with "TODO" since that implies a future action, vs when there's no TODO, it's not immediately clear to the person that this needs to be changed.

ciampo replied:
Got it! Added a comment in https://github.com/WordPress/gutenberg/pull/51097/commits/0887983f16602f12fb6a7b79730cc0ebfbedf846, let me know in case you would word it differently.

tyxla replied:
All good, thank you 🙌

tyxla commented on packages/components/src/dropdown-menu-v2/styles.ts:
```diff
@@ -23,6 +24,14 @@ const ITEM_PREFIX_WIDTH = space( 7 );
 const ITEM_PADDING_INLINE_START = space( 2 );
 const ITEM_PADDING_INLINE_END = space( 2.5 );
 
+// TODO: should bring this into the config, and make themeable
```
👍

tyxla reviewed (approved):
Works well in my testing 👍 

Code also looks good 💯 Nice work!

I've left a few questions, nothing blocking though. ✅

jasmussen commented:
This is cool, thank you for working on this. A couple of screenshots:
 
<img width="839" alt="default style" src="https://github.com/WordPress/gutenberg/assets/1204802/9c4fc944-1b2d-4bc5-a040-bbf9cd4494a1">

<img width="817" alt="toolbar style" src="https://github.com/WordPress/gutenberg/assets/1204802/6db6db7e-00ef-4d69-bb4b-10c8e6350573">

Short answer: this component is going to be such an improvement to our menus, that I would love to see a version of this land in short order. In that light, I wouldn't call _any_ of the following feedback to be blocking for the PR to land, but rather a wishlist or laundry-list of todos, weighted according to priority. First off, things I'd love to address in this PR, or shortly after, if at all possible:

First off, apply the same box shadow we have for existing popovers at the moment. That's `$shadow-popover` in sass, and `0 0.7px 1px rgba($black, 0.1), 0 1.2px 1.7px -0.2px rgba($black, 0.1), 0 2.3px 3.3px -0.5px rgba($black, 0.1)` in the code. Note that these are likely to be refined across in the near future, but would be good to just have be consistent for the time being.

Secondly, this spacing is both small and uneven. I have a sense this might be tricky to improve, but if possible we should have some larger spacing here to feel balanced with the dark outlines. I think the original design calls for 8 or 12px, but we can try 6px as well. Right now it's between 1.5px and 2px spacing, likely due to subpixel positioning issues. As noted, probably a tricky one, but anything we can do here would be good.

In the category of smaller things, again, things we can live without but nice if possible nevertheless. This icon looks to be 18x18. With almost no exceptions, our icons should always be 24x24:

<img width="269" alt="Screenshot 2023-05-31 at 09 16 42" src="https://github.com/WordPress/gutenberg/assets/1204802/51384e60-639b-4f38-8909-9d5f912c3574">

Which is curious because the check looks right:

 <img width="298" alt="Screenshot 2023-05-31 at 09 16 59" src="https://github.com/WordPress/gutenberg/assets/1204802/0abf1fcb-7a9f-4033-b26e-a002076dde56">

This helper text at 10px is fairly small and #777777. I think we use 12px helper text that should be #757575. 

Let me know if that's helpful!

ciampo committed:
Tweak box shadow to match the existing popover

ciampo committed:
Update icon size and helper text styles in Storybook example

ciampo committed:
Tweak sub menu offset

ciampo commented:
Rebased to solve conflicts and include changes from #51046.

@jasmussen:

> First off, apply the same box shadow we have for existing popovers at the moment.

Since my understanding is that we want this to look like the current dropdown/popover, I've also added a light gray border to the box shadow (as per the [current popover styles](https://github.com/WordPress/gutenberg/blob/b1202138f4e44fa6312e93f2e7d5132cf73586cd/packages/components/src/popover/style.scss#L24)). Let me know if that's ok.

> Secondly, this spacing is both small and uneven.

I assumed you're referring to the space between submenu popovers.I went ahead and I increased the gap by `4px`. 

Let me know if that's what you intended, and if the change looks good.

Also, should that gap should be different based on toolbar styles being applied, or should it be the same for all instanced of `DropdownMenu` ?

> This icon looks to be 18x18. With almost no exceptions, our icons should always be 24x24:

That 18x18 icon is just part of the Storybook example, and it's not really related to the `DropdownMenu` component.

I believe it is a decision that I took after implementing the current [design specs](https://www.figma.com/file/oituj8yH1O6z1Y6EsHY6FA/DropdownMenu?type=design&node-id=16-621). The 24x24 version of the WordPress icon felt a bit disproportionate, especially given the space between the icon and the text to its right.

I've updated the Storybook example to use a 24x24 icon, let me know what you think.

> [...]
> Which is curious because the check looks right:

The check icon is part of the `DropdownMenu` component, and I kept it at 24x24 because it felt proportionate. I don't know if there's a better term for it, but its "visual weight" at 24x24 is definitely less imposing than other icons (like the WordPress logo).

> This helper text at 10px is fairly small and #777777. I think we use 12px helper text that should be #757575.

The helper text is also just part of the Storybook example (and not part of the `DropdownMenu` component). I still went ahead and updated the styles to 12px / #757575

----

For ease of review, this is how the Storybook examples look like after the latest set of changes:

![Screenshot 2023-05-31 at 14 28 15](https://github.com/WordPress/gutenberg/assets/1083581/5e5e4139-2d96-4a60-8025-fb8f1ab62b09)
![Screenshot 2023-05-31 at 14 28 32](https://github.com/WordPress/gutenberg/assets/1083581/86f2a69f-6519-4522-8e18-69d745d76741)

ciampo committed:
add TODO comment

jasmussen commented:
Let's try 4px now, and yes if these new ones can look identical to the existing popovers (seems like that's actually the case), then we can definitely retire the old ones sooner, that'd be nice. So much we can clean up

ciampo commented:
Cool, thank you!

I'm going to merge this PR, but feel free to open a new issue with more tweaks as you see fit
